src/network/access/qnetworkaccessmanager.cpp

```diff
@@ -1652,12 +1652,23 @@ void QNetworkAccessManagerPrivate::_q_networkSessionStateChanged(QNetworkSession
     qCDebug(lcNetworkAccess) << "QNAM: network session state changed:" << state;
 
     bool reallyOnline = false;
+    bool updateStateOnly = false;
     //Do not emit the networkSessionConnected signal here, except for roaming -> connected
     //transition, otherwise it is emitted twice in a row when opening a connection.
     if (state == QNetworkSession::Connected && lastSessionState != QNetworkSession::Roaming)
         emit q->networkSessionConnected();
+
+    // This is the case when VPN resets itself and goes directly from connected back to connecting
+    // Just update session state and stop processing.
+    if (state == QNetworkSession::Connecting && lastSessionState == QNetworkSession::Connected) {
+        updateStateOnly = true;
+    }
+
     lastSessionState = state;
 
+    if (updateStateOnly)
+        return;
```

Comment:
Do we need to skip the following online checks and networkAccessibleChanged() emissions too? I suppose if a network session goes connected->connecting, the online availability might change.

Reply:
The restart itself is not being informed to ConnMan prior to the restart so that is a bit uncertain, doesn't the change on the lastSessionState trigger some event as well? That accessible just seem to have three values, unknown, accessible and not-accessible, so while in connecting state the network is accessible I think. And there is a different setup to handle the online checks here as well as the next function in this file? 

This whole thing is really such a mess. I'm not really familiar on what that networkAccessible really is for.